File: Python/0162-m-find-peak-element.py
"""
Binary Search
Time: O(logn)
Space: O(1)
"""
class Solution:
    def findPeakElement(self, nums: List[int]) -> int:
        start, end = 0, len(nums) - 1
        while start + 1 < end:
            mid = start + (end - start) // 2
            if nums[mid] < nums[mid - 1]:
                end = mid
            elif nums[mid] < nums[mid + 1]:
                start = mid
            else:
                return mid
        if nums[start] < nums[end]:
            return end
        return start

File: Java/0162-m-find-peak-element.java
/* 
Binary Search
*/
class Solution {
    public int findPeakElement(int[] nums) {
        int start = 0, end = nums.length - 1;
        int mid;
        while (start + 1 < end) {
            mid = start + (end - start) / 2;
            if (nums[mid] < nums[mid - 1])
                end = mid;
            else if (nums[mid] < nums[mid + 1])
                start = mid;
            else
                return mid;
        }
        if (nums[start] < nums[end])
            return end;
        return start;
    }
}